### PR TITLE
Update documentation for CTCLoss

### LINKED ISSUE #18249: CTCLoss documentation is unclear
## 📚 Documentation

Another student and I have been working out how to use the new CTCLoss function, and the documentation is somewhat incomplete. I want to point out some possible problems, and suggest changes to the example script.

Problems:
  - **S** is not defined for the shape of `targets`. S should be the length of the targets?
  - It appears that each value in **S** should be the index of the target, similar to CrossEntropy loss, but this is not explained
  - For the example code, why would the target indices start at 1 instead of 0?
  - `sum(target_lengths)` is not clear
  - For the example code, variables are not named in terms of **S,N,T,C**
  - It is unclear why `target_lengths` would ever need to contain values less than the true length of the target sequences, and it appears that "length" is actually the stop index? (in the example) I assume there is some intended reason for this, but it is a bit confusing why lengths are specified separately from the sequences themselves.

A possible improvement of the example script might look something like this (`x` and `y` could be replaced with `input` and `target`)

```
from torch import nn
import torch


T = 50  # Input sequence length
C = 20  # Number of classes
N = 16  # Batch size
S = 30  # Target sequence length

minimum_target_length = 10

ctc_loss = nn.CTCLoss()
x = torch.randn(T, N, C).log_softmax(2).detach().requires_grad_()    # [size] = T,N,C
y = torch.randint(1, C, (N, S), dtype=torch.long)                    # low, high, [size]

x_lengths = torch.full((N,), T, dtype=torch.long)                           # Length of inputs
y_lengths = torch.randint(minimum_target_length,S,(N,), dtype=torch.long)   # Length of targets can be variable (even if target sequences are constant length)

loss = ctc_loss(x, y, x_lengths, y_lengths)
loss.backward()

print("x.shape", x.shape)
print("y.shape", y.shape)
print("x_lengths", x_lengths)
print("y_lengths", y_lengths)

print(loss)
```

Hopefully this would save people some time figuring out the currently unstated requirements of CTCLoss.

credit to @Jacob0G for helping find some of these issues

### PR DESCRIPTION
**Pull request by rlorigro**
This is meant to resolve #18249, where I pointed out a few things that could improve the CTCLoss docs. 

My main goal was to clarify:
- Target sequences are sequences of class indices, excluding the blank index
- Lengths of `target` and `input` are needed for masking unequal length sequences, and do not necessarily = S, which is the length of the longest sequence in the batch.

I thought about Thomas's suggestion to link the distill.pub article, but I'm not sure about it. I think that should be up to y'all to decide.

I have no experience with .rst, so it might not render as expected :)

@t-vi @ezyang

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -1235,24 +1240,45 @@ class CTCLoss(_Loss):
...
     Inputs:
         log_probs: Tensor of size :math:`(T, N, C)` where `C = number of characters in alphabet including blank`,
-            `T = input length`, and `N = batch size`.
+        `T = input length`, and `N = batch size`.
             The logarithmized probabilities of the outputs
             (e.g. obtained with :func:`torch.nn.functional.log_softmax`).
-        targets: Tensor of size :math:`(N, S)` or `(sum(target_lengths))`.
-            Targets (cannot be blank). In the second form, the targets are assumed to be concatenated.
+        targets: Tensor of size :math:`(N, S)` or `(sum(target_lengths))`, where `N = batch size`, and
+        `S = maximum target length`.
```
Comment (t-vi) [inline torch/nn/modules/loss.py]: What's the effect of changing the indent here?

File: torch/nn/modules/loss.py [hunk 2]
```diff
@@ -1235,24 +1240,45 @@ class CTCLoss(_Loss):
...
+            Each element in the target sequence is a class index. Target index cannot be blank (default=0).
+            In the second form, the targets are assumed to be concatenated.
         input_lengths: Tuple or tensor of size :math:`(N)`.
-            Lengths of the inputs (must each be :math:`\leq T`)
+            Lengths of the inputs (must each be :math:`\leq T`). Lengths are specified for each sequence to achieve
+            masking. In effect, length is the stop index s for each target sequence,
+            such that :math: `target_n = targets[n,:s_n]` for each target in a batch.
         target_lengths: Tuple or tensor of size  :math:`(N)`.
-            Lengths of the targets
+            Lengths of the targets (may be :math: < S)
```
Comment (t-vi) [inline torch/nn/modules/loss.py]: Maybe *should be <= S if targets is a two-dimensional tensor* (but with `:math:`) would be clearer.

File: torch/nn/modules/loss.py [hunk 3]
```diff
@@ -1235,24 +1240,45 @@ class CTCLoss(_Loss):
...
     Example::
 
+        >>> T = 50  # Input sequence length
+        >>> C = 20  # Number of classes (excluding blank)
+        >>> N = 16  # Batch size
+        >>> S = 30  # Target sequence length of longest target in batch
+        >>>
+        >>> min_target_length = 10
+        >>> max_target_length = S
+        >>>
```
Comment (t-vi) [inline torch/nn/modules/loss.py]: I must admit that the blank lines are just too much. Also, I wonder, if we need to introduce a second name for S.

**Comment by t-vi** [comment]
Thank you for your patch! It will be nice to have the improved documentation.
I have a few minor comments. - Also did you build the documentation to see how it is rendered after your changes?

File: torch/nn/modules/loss.py [hunk 4]
```diff
@@ -1235,24 +1240,45 @@ class CTCLoss(_Loss):
...
-            `T = input length`, and `N = batch size`.
+        `T = input length`, and `N = batch size`.
             The logarithmized probabilities of the outputs
             (e.g. obtained with :func:`torch.nn.functional.log_softmax`).
-        targets: Tensor of size :math:`(N, S)` or `(sum(target_lengths))`.
-            Targets (cannot be blank). In the second form, the targets are assumed to be concatenated.
+        targets: Tensor of size :math:`(N, S)` or `(sum(target_lengths))`, where `N = batch size`, and
+        `S = maximum target length`.
+            Each element in the target sequence is a class index. Target index cannot be blank (default=0).
+            In the second form, the targets are assumed to be concatenated.
```
Comment (ramithp) [inline torch/nn/modules/loss.py]: Are the targets concatenated with or without the padding to make it a fixed (N, S)?
Maybe it would be useful to explicitly comment on that.

File: torch/nn/modules/loss.py (hunk 3 above)
Comment (rlorigro) [reply torch/nn/modules/loss.py]: I redid the naming it in the latest update. Also removed some of the self explanatory comments, and removed some spaces.

File: torch/nn/modules/loss.py (hunk 1 above)
Comment (rlorigro) [reply torch/nn/modules/loss.py]: Did you mean on line 1238?

The indent means that the remaining words in that sentence will belong to the header block when rendered. I noticed that some of the other docs don't do this, so I have actually changed it again to something else.

File: torch/nn/modules/loss.py (hunk 4 above)
Comment (rlorigro) [reply torch/nn/modules/loss.py]: I think the answer is "with padding". I actually still don't understand how the (sum(target_lengths)) form works... it doesn't seem like that actually describes a shape. Can @t-vi comment on this?

**Comment by rlorigro** [comment]
Ok I made some changes, hopefully for the better. 

Since you asked about rendering, I didn't want to try to build this because of the dependencies and errors I got when I tried, but i did find a cheap renderer online (missing the math plugin) to give a sense of the appearance

image

File: torch/nn/modules/loss.py (hunk 4 above)
Comment (t-vi) [reply torch/nn/modules/loss.py]: That's a 1-d shape. `(sum(target_lengths),)` would probably be more pythonesque.

**Review by t-vi — APPROVED** [review]
Thank you for checking the rendering! (I had half-hoped that you'd go through the trouble of installing the doc dependencies and make a habit of improving those things you find imperfect. :wink:)

From my point of view, this can go in, unless you want to do something about the less than perfect explanation of concatenated targets.

**Comment by rlorigro** [comment]
No problem. I will make one more edit tomorrow to address the lengths/concatenation, since there is no rush. Also, does the alternative concatenated form not also apply to inputs?

File: torch/nn/modules/loss.py (hunk 4 above)
Comment (rlorigro) [reply torch/nn/modules/loss.py]: Ah no I see now. Not sure why I was confused before. The comma is probably not necessary.

**Comment by t-vi** [comment]
> I will make one more edit tomorrow to address the lengths/concatenation, since there is no rush. 
Great, thanks!

> Also, does the alternative concatenated form not also apply to inputs?
No.

**Comment by ezyang** [comment]
OK, I'll wait for the last change before landing this

**Comment by rlorigro** [comment]
@ezyang @t-vi 

Ok here is the final edit. It is still less than perfect, but a lot less less than perfect than it was before. I decided to just explicitly break out the explanations for both forms of input.

This is how it renders (roughly):
image

**Comment by t-vi** [comment]
Thanks! Two quick comments:
- In the explanation of target lengths, you write <= T when it should be <= S. I missed that earlier, sorry!
- I must admit I don't understand the explanation for concatenated lengths. *If the targets are given as a 1d tensor that is the concatenation of individual targets, the target_lengths must add up to the total length of the tensor.* maybe?

**Comment by rlorigro** [comment]
Ah sorry about that, I will fix it, and use your wording for target_lengths

**Comment by rlorigro** [comment]
@t-vi @ezyang 

Documentation should be fixed now

**Comment by t-vi** [comment]
Looks good! Thanks! But there is a merge conflict for the formatting of the reduction strings.

**Comment by soumith** [comment]
I fixed the merge conflict. Thanks a lot @rlorigro !

**Comment by rlorigro** [comment]
Thanks @soumith for fixing the conflict. Does this page get updated whenever the next version of pytorch is officially released: https://pytorch.org/docs/stable/nn.html#ctcloss  ?

**Comment by soumith** [comment]
@rlorigro yes. right now master docs are already updated: https://pytorch.org/docs/master/nn.html#ctcloss

**Comment by rlorigro** [comment]
Awesome, thanks. Ah it looks like the kerning on the "math" font is making it almost unreadable in some places :(

**Comment by edchengg** [comment]
why is the input size for class is C = 20 which excluding blank?  If the network does not predict the blank in the output, why would this loss make sense...

`# Initialize random batch of input vectors, for *size = (T,N,C)
input = torch.randn(T, N, C).log_softmax(2).detach().requires_grad_()`
????

instead of

`# Initialize random batch of input vectors, for *size = (T,N,C)
input = torch.randn(T, N, C + 1).log_softmax(2).detach().requires_grad_()`

**Comment by rlorigro** [comment]
I think you are right, it should actually be `C+1` for `input` and and `C` for `target`
